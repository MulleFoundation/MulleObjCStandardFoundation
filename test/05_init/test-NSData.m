#import <MulleObjCStandardFoundation/MulleObjCStandardFoundation.h>
#include <mulle-testallocator/mulle-testallocator.h>
#include <stdio.h>
#include <stdlib.h>
#if defined(__unix__) || defined(__unix) || (defined(__APPLE__) && defined(__MACH__))
# include <unistd.h>
#endif


static int   test_i_init_with_bytes_no_copy_length_allocator_( void)
{
   NSData *obj;
   void * params_1[] =
   {
      NULL,
      "",
      "VfL Bochum",
      "Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur." /* > 256 chars */,
      "1848",
      "\xe2\x98\x84\xef\xb8\x8f\xe2\x98\x83\xef\xb8\x8f\xf0\x9f\x91\x8d\xf0\x9f\x8f\xbe" /* emoji comet, snowman, thumbs-up brown */
   };
   unsigned int   i_1;
   unsigned int   n_1 = 6;


   for( i_1 = 0; i_1 < n_1; i_1++)
   {
      @try
      {
         obj = [[[NSData alloc] mulleInitWithBytesNoCopy:params_1[ i_1]
                                                  length:params_1[ i_1]  ? strlen( params_1[ i_1]) : 0
                                               allocator:&mulle_stdlib_nofree_allocator] autorelease];
         printf( "%s\n", [obj cStringDescription]);
      }
      @catch( NSException *localException)
      {
         printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
      }
   }
   return( 0);
}


static int   test_i_init_with_bytes_no_copy_length_free_when_done_( void)
{
   NSData *obj;
   void * params_1[] =
   {
      NULL,
      "",
      "VfL Bochum",
      "Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur." /* > 256 chars */,
      "1848",
      "\xe2\x98\x84\xef\xb8\x8f\xe2\x98\x83\xef\xb8\x8f\xf0\x9f\x91\x8d\xf0\x9f\x8f\xbe" /* emoji comet, snowman, thumbs-up brown */
   };
   unsigned int   i_1;
   unsigned int   n_1 = 6;
   BOOL params_3[] =
   {
      YES,
      NO
   };
   unsigned int   i_3;
   unsigned int   n_3 = 2;
   char           *s;

   for( i_1 = 0; i_1 < n_1; i_1++)
      for( i_3 = 0; i_3 < n_3; i_3++)
      {
         @try
         {
            s = params_1[ i_1];
            if( s && params_3[ i_3])
               s = mulle_allocator_strdup( &mulle_stdlib_allocator, s);
            obj = [[[NSData alloc] initWithBytesNoCopy:s
                                                length:s ? strlen( s) : 0
                                          freeWhenDone:params_3[ i_3]] autorelease];
            printf( "%s\n", [obj cStringDescription]);
         }
         @catch( NSException *localException)
         {
            printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
         }
      }
   return( 0);
}


static int   test_i_init( void)
{
   NSData *obj;

   @try
   {
      obj = [[[NSData alloc] init] autorelease];
      printf( "%s\n", [obj cStringDescription]);
   }
   @catch( NSException *localException)
   {
      printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
   }
   return( 0);
}


static int   test_i_init_with_bytes_no_copy_length_( void)
{
   NSData *obj;
   void * params_1[] =
   {
      NULL,
      "",
      "VfL Bochum",
      "Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur." /* > 256 chars */,
      "1848",
      "\xe2\x98\x84\xef\xb8\x8f\xe2\x98\x83\xef\xb8\x8f\xf0\x9f\x91\x8d\xf0\x9f\x8f\xbe" /* emoji comet, snowman, thumbs-up brown */
   };
   unsigned int   i_1;
   unsigned int   n_1 = 6;
   char           *s;

   for( i_1 = 0; i_1 < n_1; i_1++)
   {
      @try
      {
         s = params_1[ i_1];
         if( s)
            s = mulle_allocator_strdup( &mulle_stdlib_allocator, s);
         obj = [[[NSData alloc] initWithBytesNoCopy:s
                                             length:s ? strlen( s) : 0] autorelease];
         printf( "%s\n", [obj cStringDescription]);
      }
      @catch( NSException *localException)
      {
         printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
      }
   }
   return( 0);
}


static int   test_i_init_with_data_( void)
{
   NSData *obj;
   NSData * params_1[] =
   {
      nil,
      [NSData data],
      [NSData dataWithBytes:"\xe2\x98\x84\xef\xb8\x8f\xe2\x98\x83\xef\xb8\x8f\xf0\x9f\x91\x8d\xf0\x9f\x8f\xbe" length:19] // emoji comet, snowman, thumbs-up brown
   };
   unsigned int   i_1;
   unsigned int   n_1 = 3;

   for( i_1 = 0; i_1 < n_1; i_1++)
   {
      @try
      {
         obj = [[[NSData alloc] initWithData:params_1[ i_1]] autorelease];
         printf( "%s\n", [obj cStringDescription]);
      }
      @catch( NSException *localException)
      {
         printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
      }
   }
   return( 0);
}


static int   test_i_init_with_bytes_no_copy_length_owner_( void)
{
   NSData *obj;
   void * params_1[] =
   {
      NULL,
      "",
      "VfL Bochum",
      "Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur." /* > 256 chars */,
      "1848",
      "\xe2\x98\x84\xef\xb8\x8f\xe2\x98\x83\xef\xb8\x8f\xf0\x9f\x91\x8d\xf0\x9f\x8f\xbe" /* emoji comet, snowman, thumbs-up brown */
   };
   unsigned int   i_1;
   unsigned int   n_1 = 6;
   id params_3[] =
   {
      nil,
      @"whatever",
      @1,
      @{ @"a": @1 },
      @[ @"a" ]
   };
   unsigned int   i_3;
   unsigned int   n_3 = 5;
   char           *s;

   for( i_1 = 0; i_1 < n_1; i_1++)
      for( i_3 = 0; i_3 < n_3; i_3++)
      {
         @try
         {
            s = params_1[ i_1];
            if( s)
               s = mulle_allocator_strdup( &mulle_stdlib_allocator, s);
            obj = [[[NSData alloc] mulleInitWithBytesNoCopy:s
                                                     length:s ? strlen( s) : 0
                                              sharingObject:params_3[ i_3]] autorelease];
            printf( "%s\n", [obj cStringDescription]);
            if( s)
               mulle_allocator_free( &mulle_stdlib_allocator, s);
         }
         @catch( NSException *localException)
         {
            printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
         }
      }
   return( 0);
}


static int   test_i_init_with_bytes_length_( void)
{
   NSData *obj;
   void * params_1[] =
   {
      NULL,
      "",
      "VfL Bochum",
      "Lorem ipsum dolor sit amet, consectetur adipisici elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquid ex ea commodi consequat. Quis aute iure reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur." /* > 256 chars */,
      "1848",
      "\xe2\x98\x84\xef\xb8\x8f\xe2\x98\x83\xef\xb8\x8f\xf0\x9f\x91\x8d\xf0\x9f\x8f\xbe" /* emoji comet, snowman, thumbs-up brown */
   };
   unsigned int   i_1;
   unsigned int   n_1 = 6;

   for( i_1 = 0; i_1 < n_1; i_1++)
   {
      @try
      {
         obj = [[[NSData alloc] initWithBytes:params_1[ i_1]
                                       length:params_1[ i_1] ? strlen( params_1[ i_1]) : 0] autorelease];
         printf( "%s\n", [obj cStringDescription]);
      }
      @catch( NSException *localException)
      {
         printf( "Threw a %s exception\n", [[localException name] cStringDescription]);
      }
   }
   return( 0);
}



static int   run_test( int (*f)( void), char *name)
{
   mulle_testallocator_discard();  //  w
   @autoreleasepool                //  i
   {                               //  l  l
      printf( "%s\n", name);       //  l  e  c
      if( (*f)())                  //     a  h
         return( 1);               //     k  e
   }                               //        c
   mulle_testallocator_reset();    //        k
   return( 0);
}


int   main( int argc, char *argv[])
{
   int   errors;

#ifdef __MULLE_OBJC__
   // check that no classes are "stuck"
   if( mulle_objc_global_check_universe( __MULLE_OBJC_UNIVERSENAME__) !=
         mulle_objc_universe_is_ok)
      _exit( 1);
#endif
   errors = 0;
   errors += run_test( test_i_init_with_bytes_no_copy_length_allocator_, "-initWithBytesNoCopy:length:allocator:");
   errors += run_test( test_i_init_with_bytes_no_copy_length_free_when_done_, "-initWithBytesNoCopy:length:freeWhenDone:");
   errors += run_test( test_i_init, "-init");
   errors += run_test( test_i_init_with_bytes_no_copy_length_, "-initWithBytesNoCopy:length:");
   errors += run_test( test_i_init_with_data_, "-initWithData:");
   errors += run_test( test_i_init_with_bytes_no_copy_length_owner_, "-initWithBytesNoCopy:length:owner:");
   errors += run_test( test_i_init_with_bytes_length_, "-initWithBytes:length:");

   mulle_testallocator_cancel();
   return( errors ? 1 : 0);
}
