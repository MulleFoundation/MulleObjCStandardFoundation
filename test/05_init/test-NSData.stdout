-initWithBytesNoCopy:length:allocator:
<>
<>
<56664c20 426f6368 756d>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<31383438>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
-initWithBytesNoCopy:length:freeWhenDone:
<>
<>
<>
<>
<56664c20 426f6368 756d>
<56664c20 426f6368 756d>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<31383438>
<31383438>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
-init
<>
-initWithBytesNoCopy:length:
<>
<>
<56664c20 426f6368 756d>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<31383438>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
-initWithData:
<>
<>
<e29884ef b88fe298 83efb88f f09f918d f09f8f>
-initWithBytesNoCopy:length:owner:
<>
<>
<>
<>
<>
<>
<>
<>
<>
<>
<56664c20 426f6368 756d>
<56664c20 426f6368 756d>
<56664c20 426f6368 756d>
<56664c20 426f6368 756d>
<56664c20 426f6368 756d>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<31383438>
<31383438>
<31383438>
<31383438>
<31383438>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
-initWithBytes:length:
<>
<>
<56664c20 426f6368 756d>
<4c6f7265 6d206970 73756d20 646f6c6f 72207369 7420616d 65742c20 636f6e73 65637465 74757220 61646970 69736963 6920656c 69742c20 73656420 65697573 6d6f6420 74656d70 6f722069 6e636964 756e7420 7574206c 61626f72 65206574 20646f6c 6f726520 6d61676e 6120616c 69717561 2e205574 20656e69 6d206164 206d696e 696d2076 656e6961 6d2c2071 75697320 6e6f7374 72756420 65786572 63697461 74696f6e 20756c6c 616d636f 206c6162 6f726973 206e6973 69207574 20616c69 71756964 20657820 65612063 6f6d6d6f 64692063 6f6e7365 71756174 2e205175 69732061 75746520 69757265 20726570 72656865 6e646572 69742069 6e20766f 6c757074 61746520 76656c69 74206573 73652063 696c6c75 6d20646f 6c6f7265 20657520 66756769 6174206e 756c6c61 20706172 69617475 722e>
<31383438>
<e29884ef b88fe298 83efb88f f09f918d f09f8fbe>
